tools/walletextension/vk_repository.go

```diff
@@ -0,0 +1,72 @@
+package walletextension
+
+import (
+	"fmt"
+	"github.com/ethereum/go-ethereum/common"
+	"github.com/ethereum/go-ethereum/crypto/ecies"
+)
+
+func NewViewingKeyRepository() *ViewingKeyRepository {
+	return &ViewingKeyRepository{
+		viewingKeysPrivate: make(map[common.Address]*ecies.PrivateKey),
+		viewingKeysPublic:  make(map[common.Address][]byte),
+	}
+}
+
+type ViewingKeyRepository struct {
+	viewingKeysPrivate map[common.Address]*ecies.PrivateKey // Maps an address to its private viewing key.
+	viewingKeysPublic  map[common.Address][]byte            // Maps an address to its public viewing key bytes.
+}
+
+func (v *ViewingKeyRepository) IsReady() bool {
+	return len(v.viewingKeysPrivate) > 0
+}
+
+func (v *ViewingKeyRepository) DecryptBytes(respBytes []byte) ([]byte, error) {
+	// We attempt to decrypt the response with each viewing key in turn.
+	// TODO - Avoid trying all keys for certain requests by inspecting the request body?
+	for _, privateKey := range v.viewingKeysPrivate {
+		decryptedResult, err := privateKey.Decrypt(respBytes, nil, nil)
+		if err == nil {
+			// The decryption did not error, which means we successfully decrypted the result.
+			return decryptedResult, nil
+		}
+	}
+
+	return nil, fmt.Errorf("could not decrypt the response with any of the registered viewing keys")
+}
+
+func (v *ViewingKeyRepository) suggestFromAddressForEthCall(callParams map[string]interface{}) (*common.Address, error) {
```

Comment:
feels like this doesn't belong in this vk_manager. 
let's move it to its own file, and pass the vkmanager to it.